### ISSUE TITLE
Multiple Digit Recognition (Independent fully connected layers)

### ISSUE THREAD

**Issue by ritchieng** [fchollet/keras#3928]
I'm trying to use the convolution layer as an input and to have 5 multiple fully connected layers to recognize 5 digits in the SVHN dataset. Does anyone know how to do this in Keras? I'm stuck at the at convolution layer as this branches out. I can't find anything in the documentation. 

![c53c73c8-6890-11e5-90f4-65daec69c723](https://cloud.githubusercontent.com/assets/11657655/19007025/aa2369d8-8730-11e6-9877-c5c82bdc4529.png)

```
model = Sequential()

model.add(Convolution2D(32, 3, 3, border_mode='same',
                        input_shape=(img_channels, img_rows, img_cols)))
model.add(Activation('relu'))
model.add(Convolution2D(32, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Convolution2D(64, 3, 3, border_mode='same'))
model.add(Activation('relu'))
model.add(Convolution2D(64, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

# So what happens after this to have 5 independent fully connected layers to recognise the multiple digits?
```

**Comment by crowsonkb** [comment]
For recognizing five English letters in a row given an input sound file of a word, I used the [functional API](https://keras.io/getting-started/functional-api-guide/) (`seq` is a helper function of mine):

```
def seq(*layers):
    return functools.reduce(lambda f, g: lambda x: g(f(x)), layers, lambda x: x)

model_in = Input(shape=(max_len+jitter,))
model_main = seq(
    Reshape((max_len+jitter, 1)),
    Conv1D(64, 64, subsample_length=32, activation='relu'),
    MaxPooling1D(pool_length=2),
    Flatten(),
    Dense(200, activation='relu'),
)(model_in)
branches = []
for i in range(5):
    branches.append(Dense(26, activation='softmax', name='letter_'+str(i))(model_main))
model = Model(model_in, branches)
```

**Comment by ritchieng** [comment]
@crowsonkb Thank you so much!

Would you have to make any changes to calculate accuracy? I guess the code after yours would be this? Can we still use the metric "accuracy" without any issues in Keras?

```
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(X_train, y_train, validation_data=(X_test, y_test), nb_epoch=10, batch_size=200,
    verbose=2)
model.evaluate(X_test, y_test, verbose=0)
```

**Comment by crowsonkb** [comment]
I snipped the rest because it had a lot of custom generator/callback stuff in it which wasn't necessary for the example. But yes, the next lines would be just as you described. When there are multiple outputs, the 'accuracy' metric just provides a separate accuracy score for each individual output.

**Comment by ritchieng** [comment]
@crowsonkb Would separate accuracies make sense as the final accuracy should be whether "712" matches "712" instead of accuracies of matching "7", "1" and "2"? Or does keras combine the accuracies into 1 to show the accuracy of our prediction "712" matching "712" (labelled)?

**Comment by crowsonkb** [comment]
You might have to write a custom metric to get a combined accuracy value.

**Comment by ritchieng** [comment]
@crowsonkb 

I've the following shapes (well-prepared).
X_train: (188602, 32, 32, 1)
y_train_dummy: (188602, 5, 11)
X_val: (47151, 32, 32, 1)
y_val_dummy: (47151, 5, 11)

I've the following error when I run model.fit()
`Exception: The model expects 5 input arrays, but only received one array. Found: array with shape (188602, 5, 11)` Been struggling, any idea why? I got a feeling it's my model. My data is well prepared.


```
batch_size = 128
nb_classes = 10
nb_epoch = 2

_, img_rows, img_cols, img_channels = X_train.shape

model_input = Input(shape=(img_rows, img_cols, img_channels))
x = Convolution2D(32, 3, 3, border_mode='same')(model_input)
x = Activation('relu')(x)
x = Convolution2D(32, 3, 3)(x)
x = Activation('relu')(x)
x = MaxPooling2D(pool_size=(2, 2))(x)
x = Dropout(0.25)(x)
conv_out = Flatten()(x)

x1 = Dense(nb_classes, activation='softmax')(conv_out)
x2 = Dense(nb_classes, activation='softmax')(conv_out)
x3 = Dense(nb_classes, activation='softmax')(conv_out)
x4 = Dense(nb_classes, activation='softmax')(conv_out)
x5 = Dense(nb_classes, activation='softmax')(conv_out)

lst = [x1, x2, x3, x4, x5]

model = Model(input=model_input, output=lst)

sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='categorical_crossentropy', optimizer=sgd)
model.fit(X_train, y_train_dummy, batch_size=batch_size, nb_epoch=nb_epoch, verbose=1, validation_data=(X_val, y_val_dummy))
```

**Comment by crowsonkb** [comment]
You have to supply a list of separate arrays for the five outputs, i.e. instead of the (188602, 5, 11) y_train_dummy, a list of five (188602, 11) arrays. Same for the y_val_dummy.

**Comment by ritchieng** [comment]
@crowsonkb Thanks really, it works now.

I'm struggling with accuracy because I used the default accuracy, I'm facing a very weird issue of 5 accuracies where the accuracy of the first digit is high, and the rest are low. Did you do anything in particular that addressed this? 

`Epoch 1/10 188602/188602 [==============================] - 229s - loss: 4.2374 - dense_14_loss: 1.4432 - dense_15_loss: 1.6796 - dense_16_loss: 0.9792 - dense_17_loss: 0.1342 - dense_18_loss: 0.0013 - dense_14_acc: 0.5091 - dense_15_acc: 0.3580 - dense_16_acc: 0.3715 - dense_17_acc: 0.4727 - dense_18_acc: 0.0039 - val_loss: 2.3544 - val_dense_14_loss: 0.7367 - val_dense_15_loss: 0.9298 - val_dense_16_loss: 0.5922 - val_dense_17_loss: 0.0947 - val_dense_18_loss: 0.0010 - val_dense_14_acc: 0.7791 - val_dense_15_acc: 0.6632 - val_dense_16_acc: 0.4265 - val_dense_17_acc: 0.1512 - val_dense_18_acc: 0.0154`

**Comment by jalbano** [comment]
@ritchieng Hi -- wondering if you ever came to a conclusion on the accuracy stuff -- as I'm facing the exact same problem -- I have a keras model that outputs length and 5 digits -- but really want to calc overall accuracy so that I can a) know it and b) possibly use it for a gridsearch approach. Thanks.

**Comment by shksa** [comment]
The ian goodfellow's paper has a objective P(S=s | X) = P(L=l | X) * P(S1=s1 | X) * P(S2=s2 | X) * P(S3=s3 | X)... So after you get the 10 probabilities from each classifier we need to apply the above formula to generate the probability for the entire sequence given image P(S=s | X) and then the objective function would be to maximize the log ( P(S = s | X) ).  Is this approach right and how can we do this in keras after the convolution blocks?
